Use Cardano API function `calculateMinimumUTxO` for minimum UTxO calculations.

jonathanknowles opened the pull request:
## Issue Number 

ADP-1978

## Summary

This PR uses the Cardano API function [`calculateMinimumUTxO`](https://github.com/input-output-hk/cardano-node/blob/42809ad3d420f0695eb147d4c66b90573d27cafd/cardano-api/src/Cardano/Api/Fees.hs#L1226) to replace our hand-coded [`computeMinimumAdaQuantity`](https://github.com/input-output-hk/cardano-wallet/blob/4f0b2f3c3cb9af139029418f4a2ebeb166fd0691/lib/shelley/src/Cardano/Wallet/Shelley/Compatibility/Ledger.hs#L121) function.

## Goals

1. **Ensure that we cannot underestimate minimum UTxO values.** 
    _It's imperative that we do not underestimate minimum UTxO quantities, as this may result in the creation of transactions that are unacceptable to the ledger. In the case of change generation, this would be particularly problematic, as change outputs are generated automatically, and users do not have direct control over the ada quantities generated._
    
2. **Decrease the complexity and fragility of minimum UTxO calculations within the wallet.**
    _The current design makes it trivial to specify a minimum UTxO function that's inappropriate for the era in which a transaction will be submitted. Using the wrong minimum UTxO function makes it possible to create transactions that will ultimately be rejected by the ledger._

## Design

We replace the existing `MinimumUTxOValue` type with the following pair of types:
```hs
data MinimumUTxO where
    MinimumUTxONone
        :: MinimumUTxO
    MinimumUTxOConstant
        :: Coin
        -> MinimumUTxO
    MinimumUTxOForShelleyBasedEraOf
        :: MinimumUTxOForShelleyBasedEra
        -> MinimumUTxO

data MinimumUTxOForShelleyBasedEra where
    MinimumUTxOForShelleyBasedEra
        :: ShelleyBasedEra era
        -> PParams (ShelleyLedgerEra era)
        -> MinimumUTxOForShelleyBasedEra
```
Where:
> - `MinimumUTxONone`
>     _Specifies that there is no minimum UTxO value._
> - `MinimumUTxOConstant`
>     _Specifies a constant minimum UTxO value._
> - `MinimumUTxOForShelleyBasedEra`
>     _Represents a minimum UTxO function for a Shelley-based era. Ensures that values can only be constructed with a set of protocol parameters that are specific to the given era._

Furthermore, we make the following change to `Primitive.Types.ProtocolParameters`:
```patch
-    , minimumUTxOvalue
-        :: MinimumUTxOValue
-        -- ^ The minimum UTxO value.
+    , minimumUTxO
+        :: MinimumUTxO
+        -- ^ Represents a way of calculating minimum UTxO values.
```

Finally, we introduce function [`computeMinimumCoinForUTxO`](https://github.com/input-output-hk/cardano-wallet/blob/a2c414ad3e44d1af900280b9a890bb398783e400/lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs#L51):
```hs
computeMinimumCoinForUTxO :: MinimumUTxO -> TokenMap -> Coin
```
This function computes a minimum `Coin`value for a `TokenMap` that is destined for inclusion in a transaction output.

We use the above function to re-implement [`Shelley.Transaction.constraints.txOutputMinimumAdaQuantity`](https://github.com/input-output-hk/cardano-wallet/blob/a2c414ad3e44d1af900280b9a890bb398783e400/lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs#L1500), which means that both the coin selection and migration algorithms will take advantage of this updated implementation.

## Completed QA Tasks
- [x] Transfer our existing test coverage of [`computeMinimumAdaQuantity`](https://github.com/input-output-hk/cardano-wallet/blob/4f0b2f3c3cb9af139029418f4a2ebeb166fd0691/lib/shelley/src/Cardano/Wallet/Shelley/Compatibility/Ledger.hs#L121) (including [golden tests](https://github.com/input-output-hk/cardano-wallet/blob/a2c414ad3e44d1af900280b9a890bb398783e400/lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs#L288)) to [`computeMinimumCoinForUTxO`](https://github.com/input-output-hk/cardano-wallet/blob/a2c414ad3e44d1af900280b9a890bb398783e400/lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs#L51).
- [x] Verify that our minimum UTxO calculation is stable w.r.t. to the starting ada quantity. In particular, it should obey the following property:
     - if we call [`computeMinimumCoinForUTxO`](https://github.com/input-output-hk/cardano-wallet/blob/a2c414ad3e44d1af900280b9a890bb398783e400/lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs#L51) and receive `Coin` value `c1`
     - if we call the Cardano API function [`calculateMinimumUTxO`](https://github.com/input-output-hk/cardano-node/blob/42809ad3d420f0695eb147d4c66b90573d27cafd/cardano-api/src/Cardano/Api/Fees.hs#L1226) with `Coin` value `c1` and receive `Coin` value `c2`
     - then `c2` **_must not_** be greater than `c1`.
   (See [`prop_computeMinimumCoinForUTxO_shelleyBasedEra_stability`](https://github.com/input-output-hk/cardano-wallet/blob/a2c414ad3e44d1af900280b9a890bb398783e400/lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs#L213).)
- [x] Determine what level of test coverage is applied to the Cardano API function [`calculateMinimumUTxO`](https://github.com/input-output-hk/cardano-node/blob/42809ad3d420f0695eb147d4c66b90573d27cafd/cardano-api/src/Cardano/Api/Fees.hs#L1226). Ensure that we build appropriate test coverage to cover any deficits.
- [x] Ensure that we have test coverage to cover all valid lengths of addresses.
- [x] Ensure that we have test coverage to characterize the effect of different coin sizes (up to the limit of `txOutMaxCoin`).
- [x] Ensure that we have test coverage to characterize the effect of different token quantities (up to the limit of `txOutMaxTokenQuantity`).

## Completed Implementation Tasks

- [x] Determine the most appropriate way to initialize field `minimumUTxO` within function `fromBlockfrostPP`. See 3c035604572b3a0805864d5493ebc4a409f391cb and ADP-1994.
- [x] Remove type `MinimumUTxOValue` from `Primitive.Types`.
- [x] Remove function [`computeMinimumAdaQuantity`](https://github.com/input-output-hk/cardano-wallet/blob/4f0b2f3c3cb9af139029418f4a2ebeb166fd0691/lib/shelley/src/Cardano/Wallet/Shelley/Compatibility/Ledger.hs#L121) and associated tests.

## Completed Admin Tasks 

- [x] Create a ticket to review and revise the behaviour of the Blockfrost protocol parameter conversion function so that it does not hard-code the choice of minimum UTxO function to the Alonzo era (which uses a coins-per-word calculation). See ADP-1994.
- [x] Create an issue on `cardano-node` to report the fact that the [`calculateMinimumUTxO`](https://github.com/input-output-hk/cardano-node/blob/42809ad3d420f0695eb147d4c66b90573d27cafd/cardano-api/src/Cardano/Api/Fees.hs#L1226) function does not appear to reach a fixed point before returning its result. (See https://github.com/input-output-hk/cardano-node/issues/4163). The current behaviour appears to be as follows:
    - If we call `calculateMinimumUTxO` with `Coin 0` and get `Coin a` back.
    - If we call `calculateMinimumUTxO` with `Coin a` and get `Coin b` back.
    - It's possible for `b` to be greater than `a`.

jonathanknowles committed:
Miscellaneous formatting fixes.

jonathanknowles committed:
Add function `fromAllegraPParams`.

jonathanknowles committed:
Add function `fromMaryPParams`.

jonathanknowles committed:
Specialize function `fromShelleyPParams` to the Shelley era.

jonathanknowles committed:
Add type `MinimumUTxO` with generators and shrinkers.

jonathanknowles committed:
Add field `minimumUTxO` to `ProtocolParameters` primitive type.

jonathanknowles committed:
Add function `computeMinimumCoinForUTxO`.

jonathanknowles committed:
Use `computeMinimumCoinForUTxO` within `TransactionLayer.constraints`.

jonathanknowles committed:
Simplify minimum UTxO value lookup within `toApiNetworkParameters`.

jonathanknowles committed:
Remove field `minimumUTxOvalue` from primitive type `ProtocolParameters`.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs:
```diff
@@ -884,6 +881,8 @@ fromBlockfrostPP network BF.ProtocolParams{..} = do
         , maximumCollateralInputCount = maxCollateralInputs
         , minimumCollateralPercentage = collateralPercent
+        -- TODO: Surely this is not right:
+        , minimumUTxO = MinimumUTxONone
```
I've added a task to the PR description to determine the correct way to initialize this field.

jonathanknowles replied:
Update: in a discussion with @Unisay, we decided to make this PR preserve the existing behaviour, so that the Blockfrost network layer continues to specify the Alonzo-specific coins-per-UTxO-word-based function. The only effect of applying this PR should be that the choice will become more explicit and obvious.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs:
```diff
@@ -0,0 +1,57 @@
+
+prop_computeMinimumCoinForUTxO :: MinimumUTxO -> TokenMap -> Property
+prop_computeMinimumCoinForUTxO minimumUTxO m = property $
+    computeMinimumCoinForUTxO minimumUTxO m >= Coin 0
```
This is just a placeholder property. More properties will be added later. (See PR description.)

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs:
```diff
@@ -0,0 +1,101 @@
+        -- TODO:
+        -- Use a suitable length that is at least as long as any valid address:
+        dummyAddress :: Address
+        dummyAddress = Address $ BS.pack $ 97 : replicate 28 0
```
Reminder: replace these magic numbers (`97` and `28`) with named identifiers and/or add explanations for their provenance.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2167,7 +2168,7 @@ mockProtocolParameters = dummyProtocolParameters
         , getMaxExecutionUnits = ExecutionUnits 10_000_000_000 14_000_000
         }
-    , minimumUTxOvalue = MinimumUTxOValue $ Coin 1000000
+    , minimumUTxO = MinimumUTxOConstant $ Coin 1000000
```
Reminder: use `NumericUnderscores` here.
```suggestion
    , minimumUTxO = MinimumUTxOConstant $ Coin 1_000_000
```

jonathanknowles committed:
Add dependency on library `int-cast`.

jonathanknowles committed:
Add dependency on library `generics-sop`.

jonathanknowles committed:
Use type name `MinimumUTxOForShelleyBasedEra`.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs:
```diff
@@ -0,0 +1,166 @@
+
+instance Arbitrary TokenMap where
+    arbitrary = genTokenMap
+    shrink = shrinkTokenMap
```
I think this instance is no longer used. If so, we can delete it.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs:
```diff
@@ -0,0 +1,155 @@
+        maximumAddressLength = 57
+
+    dummyCoin :: Coin
+    dummyCoin = Coin $ intCast @Word64 @Natural $ maxBound
```
Did any smaller values cause the property test to fail? 😃 (both re coin and addr)

Anviking replied:
Also, I thought the address needed to be valid 🤔 This works?

jonathanknowles replied:
> Did any smaller values cause the property test to fail? smiley (both re coin and addr)

For addresses, a shorter dummy address does indeed cause the property to fail:

```hs
  lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs:90:9: 
  1) computeMinimumCoinForUTxO prop_computeMinimumCoinForUTxO_shelleyBasedEra_bounds
       Falsified (after 5 tests and 2 shrinks):
         TokenBundle {coin = Coin 0, tokens = TokenMap (fromList [])}
         ShelleyAddress Mainnet (KeyHashObj (KeyHash "799949ef8b7aad3c1753f658b190be1a919f5057f8ea8dd187b3a1de")) (StakeRefBase (KeyHashObj (KeyHash "dc3c5f13c7dec0337e4c53de6e675d3be50e6788d9d91ada26e10ffc")))
         ShelleyBasedEraBabbage ProtocolParameters {protocolParamProtocolVersion = (7,0), protocolParamDecentralization = Nothing, protocolParamExtraPraosEntropy = Nothing, protocolParamMaxBlockHeaderSize = 0, protocolParamMaxBlockBodySize = 0, protocolParamMaxTxSize = 2048, protocolParamTxFeeFixed = 0, protocolParamTxFeePerByte = 0, protocolParamMinUTxOValue = Nothing, protocolParamStakeAddressDeposit = Lovelace 0, protocolParamStakePoolDeposit = Lovelace 0, protocolParamMinPoolCost = Lovelace 0, protocolParamPoolRetireMaxEpoch = EpochNo 0, protocolParamStakePoolTargetNum = 100, protocolParamPoolPledgeInfluence = 0 % 1, protocolParamMonetaryExpansion = 0 % 1, protocolParamTreasuryCut = 0 % 1, protocolParamUTxOCostPerWord = Just (Lovelace 55999944), protocolParamCostModels = fromList [], protocolParamPrices = Just (ExecutionUnitPrices {priceExecutionSteps = 0 % 1, priceExecutionMemory = 0 % 1}), protocolParamMaxTxExUnits = Just (ExecutionUnits {executionSteps = 0, executionMemory = 0}), protocolParamMaxBlockExUnits = Just (ExecutionUnits {executionSteps = 0, executionMemory = 0}), protocolParamMaxValueSize = Just 0, protocolParamCollateralPercent = Just 150, protocolParamMaxCollateralInputs = Just 5}
         BS.length (unAddress maxLengthAddress)):
             29
         
         BS.length (unAddress (fromCardanoAddress address)):
             57
         
         BS.length (Cardano.serialiseToRawBytes address)):
             57
         
         ourResult:
             Coin 1420998579
         
         apiResultMaxBound:
             Coin 1420998579
         
         apiResultMinBound:
             Coin 1560998439
         
         Condition violated: ourResult >= apiResultMinBound

  To rerun use: --match "/computeMinimumCoinForUTxO/prop_computeMinimumCoinForUTxO_shelleyBasedEra_bounds/"
```

jonathanknowles replied:
> Also, I thought the address needed to be valid thinking This works?

Choosing an invalid address as the dummy address does indeed cause a **"malformed address"** error, thrown by our `toCardanoTxOut` function:
```hs
 Exception thrown while showing test case:
   INTERNAL ERROR: toCardanoTxOut: malformed address
   CallStack (from HasCallStack):
     error, called at lib/core/src/Cardano/Wallet/Util.hs:78:21 in main:Cardano.Wallet.Util
     internalError, called at lib/core/src/Cardano/Wallet/Util.hs:86:23 in main:Cardano.Wallet.Util
     tina, called at lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:1901:21 in main:Cardano.Wallet.Shelley.Compatibility
     toAlonzoTxOut, called at lib/shelley/src/Cardano/Wallet/Shelley/Compatibility.hs:1833:31 in main:Cardano.Wallet.Shelley.Compatibility
```

This apparently is a "valid" address (it's accepted by both our `toCardanoTxOut` function and the Cardano API `calculateMinimumUTxO` function):
```hs
maxLengthAddress :: Address
maxLengthAddress =
    Address $ BS.pack $ replicate maxAddressLength nullByte
  where
    maxAddressLength :: Int
    maxAddressLength = 57

    nullByte :: Word8
    nullByte = 0
```

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs:
```diff
@@ -0,0 +1,166 @@
+
+prop_computeMinimumCoinForUTxO_shelleyBasedEra_lowerBound
+    :: TokenBundle
+    -> Cardano.Address Cardano.ShelleyAddr
```
What about byron addresses? (I'm slightly worried, or at least certainly curious about these)

jonathanknowles replied:
> What about byron addresses? (I'm slightly worried, or at least certainly curious about these)

I'll investigate, and report back.

jonathanknowles replied:
> What about byron addresses? (I'm slightly worried, or at least certainly curious about these)

I've revised the property to also test with Byron addresses. See 91217999a78e6b9ff2cd3ef1206435865587d4b9.

It currently still passes.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network/Node.hs:
```diff
@@ -714,9 +716,9 @@ mkTipSyncClient tr np onPParamsUpdate onInterpreterUpdate onEraUpdate = do
-                (fromShelleyPParams eraBounds ppNode
+                (fromAllegraPParams eraBounds ppNode
                     <$> LSQry Shelley.GetCurrentPParams)
-                (fromShelleyPParams eraBounds ppNode
```
Is there no simple way to keep using `fromShelleyPPararams` for Allegra and Mary rather than duplicating it?

I believe large amounts of duplication in "compatibility layers" has been a cause of bugs in the past. Though if we are moving to embrace cardano-api types soon anyway, it might not matter _too_ much...

...But shouldn't adding some `IsShelleyBasedEra era` constraint to `fromShelleyPParams` and then constructing the minimumUTxOValue there work?

jonathanknowles replied:
> Is there no simple way to keep using fromShelleyPPararams for Allegra and Mary rather than duplicating it?

This would also be my preference!

> ...But shouldn't adding some `IsShelleyBasedEra era` constraint to `fromShelleyPParams` and then constructing the minimumUTxOValue there work?

I did try this, but was unable to get the types to line up (though I can't remember the precise error). I'll give it another try!

jonathanknowles replied:
Hi @Anviking 

Just a thought:

Perhaps it's clearer as it currently is, with one separate function for each era?

Given that these functions are always called with a particular era in hand, it's hard for people to use the wrong function for a given era, since every function requires an era-specific `PParams` value.

Of course, we could try to merge `from{Shelley,Allegra,Mary}PParams` into one function, but what would we call the function?

* We could call it `fromShelleyBasedEraPParams`, but this name implies that it ought to be usable for any Shelley-based era, including those after `Mary`, which would be a mistake.
* We could call it `fromShelleyAllegraMaryPParams`, and this name would _perhaps_ be more accurate, but it seems rather awkward.

jonathanknowles committed:
Extract out constants `maxLengthCoin` and `maxLengthAddress`.

jonathanknowles committed:
Revise `computeMinimumCoinForUTxO` property to check upper and lower bounds.

jonathanknowles committed:
Add function `genAddressAny` to `Cardano.Api.Gen`.

jonathanknowles committed:
Revise `computeMinimumCoinForUTxO` property to test Byron and Shelley addresses.

jonathanknowles committed:
Add explanatory comments to properties in `MinimumUTxOSpec`.

jonathanknowles committed:
Use more robust generator for `TokenBundle` values.

jonathanknowles committed:
Use coins-per-word-based calculation for Blockfrost.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs:
```diff
@@ -832,10 +838,9 @@ fromBlockfrostPP network BF.ProtocolParams{..} = do
-            <?#> "CoinsPerUtxoWord"
+    coinsPerUTxOWord <- Ledger.Coin
+        <$> intCast @_ @Integer _protocolParamsCoinsPerUtxoWord
+        <?#> "CoinsPerUtxoWord"
```
Suggestion: directly construct the `MinimumUTxO` value here, rather than creating this temporary value that will only be consumed once.

jonathanknowles replied:
Fixed in 3c035604572b3a0805864d5493ebc4a409f391cb.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs:
```diff
@@ -0,0 +1,172 @@
+        Left e ->
+            -- We assume that the provided protocol parameters record has all
+            -- the required parameters for the given era. If this assumption is
+            -- violated, we have no way to continue, and must raise an error:
```
Suggestion: explain why this should be impossible. (The `MinimumUTxOForShelleyBasedEra` constructor already requires an era-specific protocol parameters object, which should mean that missing fields are impossible.)

jonathanknowles replied:
Fixed in a2c414ad3e44d1af900280b9a890bb398783e400.

jonathanknowles committed:
Extract out `getMinimumUTxOFunction` within Blockfrost network layer.

This function encapsulates the logic of constructing an era-specific
minimum UTxO function, and reduces the complexity of the
`fromBlockfrostPP` function.

jonathanknowles committed:
Generalize function `verify` in `Test.QuickCheck.Extra`.

This allows `verify` to modify any `Testable` value.

jonathanknowles committed:
Add property `prop_computeMinimumCoinForUTxO_shelleyBasedEra_stability`.

This function compares the stability of:

- the Cardano API function 'calculateMinimumUTxO'
- the wallet function 'computeMinimumCoinForUTxO'

jonathanknowles committed:
Simplify property `prop_computeMinimumCoinForUTxO_shelleyBasedEra_bounds`.

Simplify this property by moving the error pattern match to inner
function `apiComputeMinCoin`. We assume that error pattern will never
match. But if it does, then this property will still fail.

jonathanknowles committed:
Add golden tests for minimum UTxO values.

We add golden minimum UTxO values for all Shelley-based eras.

jonathanknowles committed:
Include golden `TokenMap` values in generator for `TokenMap`.

jonathanknowles committed:
Disable `PartialTypeSignatures` in `Shelley.Compatibility`.

This extension was only used once in the whole module.

In the particular case it was used, it is arguably clearer and safer to
state the era explicitly.

jonathanknowles committed:
Remove legacy function `computeMinimumAdaQuantity` from `Compatibility.Ledger`.

jonathanknowles committed:
Remove legacy type `MinimumUTxOValue` from `Primitive.Types`.

jonathanknowles committed:
Improve clarity of counterexample output within golden tests.

In particular, we indicate more clearly which value was expected, and
which value was returned.

For example:

```
Failures:

  lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs:490:9:
  1) computeMinimumCoinForUTxO, Golden Tests, goldenTests_computeMinimumCoinForUTxO Babbage, golden test #3
       Falsified (after 1 test):
         resultExpected:                                                                                                                                                                           Coin 1323170
             Coin 1323170

         resultReturned:
             Coin 1357650

         Condition violated: resultReturned == resultExpected
```

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs:
```diff
@@ -0,0 +1,533 @@
+goldenMinimumUTxO_Babbage =
+    minimumUTxOForShelleyBasedEra ShelleyBasedEraBabbage
+        -- Value derived from 'mainnet-alonzo-genesis.json':
+        def {Babbage._coinsPerUTxOByte = Ledger.Coin 4_310}
```
Suggestion: explain exactly how this value was derived from `mainnet-alonzo-genesis.json`.

(Answer: by taking `coinsPerUTxOWord = 34482` and dividing it by 8, rounding down).

jonathanknowles replied:
Fixed in a2c414ad3e44d1af900280b9a890bb398783e400.

jonathanknowles committed:
Add further explanatory comments.

In response to review feedback:
https://github.com/input-output-hk/cardano-wallet/pull/3368#discussion_r913636498
https://github.com/input-output-hk/cardano-wallet/pull/3368#discussion_r918476952

jonathanknowles committed:
Add commentary to `prop_computeMinimumCoinForUTxO_shelleyBasedEra_stability`.

We further explain the purpose of this property.

If the Cardano API function `calculateMinimumUTxO` is ever changed so
that it computes a fixed point before returning its result, this
property will fail.

However, this is valuable information, as it tells us that we might be
able to revise our implementation of `computeMinimumCoinForUTxO` to
reduce the level of overestimation.

In response to review feedback.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/MinimumUTxO/Gen.hs:
```diff
@@ -0,0 +1,128 @@
+
+    genMinimumUTxOConstant :: Gen MinimumUTxO
+    genMinimumUTxOConstant = MinimumUTxOConstant . Coin
+        <$> genInterestingCoinValue
```
You could also use `genCoin` from `Cardano.Api.Gen`

jonathanknowles replied:
Hi @Anviking

I had a look for `genCoin` within `Cardano.Api.Gen`, but could only find `genLovelace`. I'm guessing you mean this function?

According to the current definition of `genLovelace`, it has a chance of generating values close to `txOutMaxLovelace`:

```hs
genLovelace :: Gen Lovelace
genLovelace = frequency
    [ (10, Lovelace . intCast . getNonNegative @Int <$> arbitrary)
    , (50, choose (1_000_000, 1_000_000_000))
    , (10, choose (txOutMinLovelace, txOutMaxLovelace))
    , (30, genEncodingBoundaryLovelace)
    ]
```

But for `genMinimumUTxO`, the values we generate will ultimately be multiplied by the number of bytes (or words) in the serialized representation of a `TxOut` in order to calculate the minimum ada quantity. So if we use the `genLovelace` definition above, we could end up computing minimum ada quantities that are greater than `txOutMaxLovelace`.

jonathanknowles replied:
Hi @Anviking 

After reading your comment, I spent a bit of time thinking about the existing `genInterestingLedgerCoin` generator.

One problem with this generator is that we reuse it for all eras within `genMinimumUTxO`. But the `Coin` value has a different meaning in each era, and realistic values have different magnitudes in different eras.

For example, on average, we would expect these values to have different magnitudes from one another:
   - the `coinsPerUTxOWord` parameter (Alonzo era)
   - the `coinsPerUTxOByte` parameter (Babbage era)

With this in mind, I think it would be better to generate values that are appropriate to the era. And to make them more realistic, we can generate values that are of a similar magnitude to real-life `mainnet` protocol parameters.

I've made this change in 114cf90e0b8be67b730b9c296f31f2a49a5ec4d8.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs:
```diff
@@ -0,0 +1,188 @@
+-- not have direct control over the ada quantities generated.
+--
+-- However, while we cannot underestimate minimum UTxO quantities, we are at
+-- liberty to moderately overestimate them.
```
Of course, overestimation is a problem when we ourselves validate that outputs satisfy the minUTxOValue, such as in balanceTx (indirectly via coin-selection), or constructTx or even the old payments endpoint. (— I believe this is the case)

jonathanknowles replied:
> Of course, overestimation is a problem when we ourselves validate that outputs satisfy the minUTxOValue, such as in balanceTx.

Hi @Anviking 

Are you able to give an example of how slight overestimation would create a problem with `balanceTx`? I'd definitely like to examine this further.

Anviking replied:
Say you provide the following partial tx to balanceTx:
```
emptyTx { outputs = [(shortPointerAddr, 0.99 ada)] }
```

The ada quantity might be valid to the ledger (`calculateMinimumUTxO (shortPointerAddr, 0.99 ada) >= 0.99`), but what if the wallet overestimates the requirement and requires `>= 1.0 ada`? 

Then I _believe_ coin-selection would reject the pre-selected output here: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/CoinSelection/Internal/Balance.hs#L947-L964 (I haven't tested though)

And I believe the same should apply for constructTx or the normal payments endpoint.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/MinimumUTxO.hs:
```diff
@@ -0,0 +1,188 @@
+-- liberty to moderately overestimate them.
+--
+-- Since the minimum UTxO quantity function is monotonically increasing in the
+-- serialized length of its input, if we supply a 'TxOut' with an address and
```
> monotonically increasing in the serialized length of its input,

Are you 100% sure about this? I.e. are you 100% sure there are no exotic pairs of pre-babbage outputs such that
```
serializedLength out1 < serializedLength out2 && minUTxOValue out1 > minUTxOValue out2
```

Perhaps replacing "serialized length of its input" with "serialised length of the ada quantity and address" might be what we really mean.

jonathanknowles replied:
Ah that's a good point, thanks for raising it!

I'll adjust this comment as you suggest.

jonathanknowles replied:
@Anviking Fixed in eb446c64aa6febf3fd3f06f950ef904f2ec019c4.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Network/Blockfrost.hs:
```diff
@@ -930,6 +934,33 @@ fromBlockfrostPP network BF.ProtocolParams{..} = do
+    minimumUTxOForAlonzoEra :: Ledger.Coin -> MinimumUTxO
+    minimumUTxOForAlonzoEra coinsPerUTxOWord = minimumUTxOForShelleyBasedEra
+        ShelleyBasedEraAlonzo
+        def {Alonzo._coinsPerUTxOWord = coinsPerUTxOWord}
```
Using `def { }` in production seems a bit sketchy, but if it's just for the time being then sure.

jonathanknowles replied:
@Anviking I agree with you that it's definitely non-ideal to use `def` here. 👍🏻 

@Unisay and I had a discussion about this last week, and we came to the conclusion that in this case, it would be acceptable to use a non-ideal workaround (which includes hard-coding the calculation to the Alonzo-era calculation), on the basis that the Blockfrost integration is still a work-in-progress feature. We have a ticket to change this after the Vasil HF here: ADP-1994

While I also dislike `def`, I think it highlights another issue: the Cardano API function `calculateMinimumUTxO` (and other similar functions) requires us to supply a fairly large set of parameters, but in the end it only uses **_one_** of these parameters. 😄 

So even though we know we have the exact parameter it needs, we still need to find some other way to supply values for the other parameters, even though we know it will never use those other values.

I would be interested in exploring designs that allow `calculateMinimumUTxO` to take exactly the parameter(s) it needs (in an era-dependent way), and nothing else.

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs:
```diff
@@ -0,0 +1,542 @@
+    , (goldenTokenMap_4, Coin 1_999_998)
+    ]
+
+goldenMinimumCoins_Alonzo :: [(TokenMap, Coin)]
+goldenMinimumCoins_Alonzo =
+    [ (goldenTokenMap_0, Coin   999_978)
+    , (goldenTokenMap_1, Coin 1_344_798)
+    , (goldenTokenMap_2, Coin 1_448_244)
+    , (goldenTokenMap_3, Coin 1_620_654)
+    , (goldenTokenMap_4, Coin 1_862_028)
+    ]
+
+goldenMinimumCoins_Babbage :: [(TokenMap, Coin)]
+goldenMinimumCoins_Babbage =
+    [ (goldenTokenMap_0, Coin   995_610)
+    , (goldenTokenMap_1, Coin 1_150_770)
+    , (goldenTokenMap_2, Coin 1_323_170)
+    , (goldenTokenMap_3, Coin 1_323_170)
+    , (goldenTokenMap_4, Coin 2_012_770)
+    ]
```
👍

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types/MinimumUTxO.hs:
```diff
@@ -0,0 +1,142 @@
+    MinimumUTxOConstant
+        :: Coin
+        -> MinimumUTxO
+        -- ^ Indicates a constant minimum UTxO value.
```
You could mention that this is only used for testing.

jonathanknowles replied:
> You could mention that this is only used for testing.

Thanks for this suggestion! I'll add a comment.

jonathanknowles replied:
@Anviking Fixed in 727c7cb95840f0deb99b2570e4fc7df6c8211219.

jonathanknowles committed:
Improve realism of generated test protocol parameters in `MinimumUTxO.Gen`.

This commit introduces generator function `genLedgerCoinOfSimilarMagnitude`,
which, when given a coin value, will generate another coin value that is
of a similar magnitude.

When generating test values of `MinimumUTxO`, we apply the generator
function `genLedgerCoinOfSimilarMagnitude` to protocol parameter values
obtained from real mainnet genesis files. This enables us to generate
more realistic values of `MinimumUTxO`.

In response to review feedback:
https://github.com/input-output-hk/cardano-wallet/pull/3368#discussion_r918912368

jonathanknowles committed:
Reuse test protocol parameters from `MinimumUTxO.Gen` in `MinimumUTxOSpec`.

Since we've already defined a set of test protocol parameter values
within `MinimumUTxO.Gen`, we might as well re-use these test parameter
values within `MinimumUTxOSpec`. This reduces the amount of repetition.

jonathanknowles committed:
Adjust monotonicity comment for function `embedTokenMapWithinPaddedTxOut`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3368#discussion_r918940878

jonathanknowles committed:
Describe motivation for `MinimumUTxOConstant` constructor.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/3368#discussion_r918974435
